keps/sig-storage/20190204-scale-testing.md

```diff
@@ -80,11 +83,47 @@ Adding storage scale and performance tests will help:
 
 ## Proposal
 
-### WIP Test Environment
-
-The tests should be developed and run in sig-scalability’s test framework and infrastructure.
-
-TODO details and links
+### Test Framework
+
+The tests should be developed and run in sig-scalability’s test
+[Cluster Loader framework](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/README.md)
+and infrastructure.
+
+The framework already supports measurements for Pod startup latency and
+scheduling metrics. More measurements can be added to get finer-grained details
+for volume operations:
+
+* Gather [volume operation
+  metrics](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/metrics.go)
+  from kube-controller-manager
+  * Caveat: These metrics measure the time of a single operation attempt and not
+    the e2e time across retries. These metrics may not be suitable to measure
+    latency from a pod's perspective.
+* Measure time starting from pod scheduling to volume operation completion times
+  for provision, attach and mount. This requires changes in Kubernetes to add
+  timestamps in PersistentVolumeClaim, VolumeAttachment, and Pod statuses,
+  respectively.
+  * TODO: how to measure time reliably for deletion events, where the API object
+    deletion is the trigger?
```

Comment:
I'm not sure whether relying on the API to expose internals is the right approach. It probably isn't going to get us all the way.

As it has been pointed out in the review of the tracing proposal, mapping to spans (whether in OpenTracing or OpenCensus) doesn't look like the right model either.

Perhaps we really just need an event collection mechanism that is independent of the Kubernetes API, and then do some data analysis on that?

Reply:
I agree that extending the API doesn't sound like a good option.
Though, it's possible to e.g. consider setting an annotation to expose that information - there is already a precedent for it: https://github.com/kubernetes/kubernetes/pull/69946 though we should be careful with things like this.

Reply:
I'm not convinced that adding timestamps of important events to api objects is exposing internal details.  I was picturing this after looking at how the pod startup sli is calculated, and seeing that it looks like Pod container's start timestamp.

Reply:
Let me rephrase my thoughts (also, I don't have full insight in what exactly you're suggesting, so this may not make sense - let me knwo if that's the case).
I guess that I was trying to say is that, I would be reluctant to extending the overall Pod api, just to enable tracking of some internal metrics. If what you have on your mind is appropriately setting timestamps e.g. in Pod conditions or sth like that, I think that can make sense (though I would need to understand a bit better what exactly you're proposing).